### PR TITLE
Introduce semantic query cache

### PR DESCRIPTION
**Pull request by kasper-piskorski**
# Why is this PR needed?
Ticket: #3441 
# What does the PR do?
Introduces the following concepts:
- Query Subsumption:
```
Subsumption relation between a query C (child) and a provided query P (parent) holds if:

C <= P,

is true, meaning that P is more general than C (C specialises P) 
and their respective answer sets meet:
 
answers(C) subsetOf answers(P)
    
i. e. the set of answers of C is a subset of the set of answers of P.
```
- Semantic Difference 
```
Semantic difference between query C and P defines a specialisation operation
required to transform query P into a query equivalent to C.
```

Those concepts form the basis of the operation of the semantic cache: if we are looking for answers to query C and the cache already contains query P which has all DB entries (db-complete), we can propagate the answers, possibly specialising them (applying the semantic difference).
# Does it break backwards compatibility?
No.
# List of future improvements not on this PR
Make it transaction-level.

### CODE REVIEW

**Comment by haikalpribadi** [comment]
This PR now fails to merge because of this change @kasper-piskorski https://github.com/graknlabs/grakn/pull/4634

File: server/src/graql/internal/reasoner/ResolutionIterator.java [hunk 1]
```diff
@@ -49,7 +47,7 @@
     private final ReasonerQueryImpl query;
     private final Set<ConceptMap> answers = new HashSet<>();
 
-    private final SimpleQueryCache<ReasonerAtomicQuery> cache = new SimpleQueryCache<>();
+    private final IndexedSemanticCache cache = new IndexedSemanticCache();
```
Comment (flyingsilverfin) [inline server/src/graql/internal/reasoner/ResolutionIterator.java]: Rename `Indexed` to `Multilevel`

File: server/src/graql/internal/reasoner/cache/Index.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+package grakn.core.graql.internal.reasoner.cache;
+
+import com.google.common.collect.ImmutableSet;
+import grakn.core.graql.Var;
+import java.util.HashSet;
+import java.util.Objects;
+import java.util.Set;
+
+/**
+ * TODO
```
Comment (flyingsilverfin) [inline server/src/graql/internal/reasoner/cache/Index.java]: what's this TODO supposed to be?

File: server/src/graql/internal/reasoner/cache/IndexedAnswerSet.java [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+import grakn.core.graql.answer.ConceptMap;
+import grakn.core.graql.query.answer.ConceptMapImpl;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Set;
+import java.util.stream.Stream;
+
+/**
+ * TODO
```
Comment (flyingsilverfin) [inline server/src/graql/internal/reasoner/cache/IndexedAnswerSet.java]: Same documentation missing

**Pull request merged**